corehq/motech/auth.py

```diff
@@ -39,3 +105,153 @@ def __call__(self, r):
+
+        resource_owner_key = self.last_token['oauth_token']
+        resource_owner_secret = self.last_token['oauth_token_secret']
+        return OAuth1(
+            self.client_id,
+            client_secret=self.client_secret,
+            resource_owner_key=resource_owner_key,
+            resource_owner_secret=resource_owner_secret
+        )
+
+
+class BearerAuthManager(AuthManager):
+    """
+    Like OAuth 2.0 Password Grant, but doesn't use a client ID or
+    client secret.
+    """
+
+    def __init__(self, username, password):
+        self.username = username
+        self.password = password
+
+    def get_auth(self):
+        return HTTPBearerAuth(self.username, self.password)
+
+
+class OAuth2PasswordGrantManager(AuthManager):
+    """
+    Follows the OAuth 2.0 resource owner password credentials (aka
+    password) grant type flow.
+    """
+
+    def __init__(
+        self,
+        base_url: str,
+        username: str,
+        password: str,
+        client_id: str,
+        client_secret: str,
+        api_settings: OAuth2ApiSettings,
+        last_token: Optional[dict] = None,
+    ):
+        self.base_url = base_url
+        self.username = username
+        self.password = password
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.api_settings = api_settings
+        self.last_token = last_token
+
+    def get_session(self):
+
+        def set_last_token(token):
+            # Used by OAuth2Session
+            self.last_token = token
+
+        if not self.last_token:
+            client = LegacyApplicationClient(client_id=self.client_id)
+            session = OAuth2Session(client=client)
+            token_url = self.get_url(self.api_settings.token_url)
+            if self.api_settings.pass_credentials_in_header:
+                auth = HTTPBasicAuth(self.client_id, self.client_secret)
+                self.last_token = session.fetch_token(
+                    token_url=token_url,
+                    username=self.username,
+                    password=self.password,
+                    auth=auth,
+                )
+            else:
+                self.last_token = session.fetch_token(
+                    token_url=token_url,
+                    username=self.username,
+                    password=self.password,
+                    client_id=self.client_id,
+                    client_secret=self.client_secret,
+                )
+
+        # Return session that refreshes token automatically
+        refresh_url = self.get_url(self.api_settings.refresh_url)
+        refresh_kwargs = {
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+        }
+        return OAuth2Session(
+            self.client_id,
+            token=self.last_token,
+            auto_refresh_url=refresh_url,
+            auto_refresh_kwargs=refresh_kwargs,
+            token_updater=set_last_token
+        )
+
+    def get_url(self, uri):
```

Comment:
should this be a private method?

Reply:
The `Requests` class has an almost identical method. I've pulled them out into a function. I'm not a big fan of "utils.py" files, but this one was already there, and seemed like the most reasonable place to put it. 60159bb